digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/AwsService.scala

```diff
@@ -0,0 +1,299 @@
+          case None =>
+            obstacleService.create(newObstacle, AwsUser, link.get)
+        }
+    }
+  }
+
+  private def updateRoadlink(properties: Map[String, Any], links: Seq[RoadLink]) = {
+    val speedLimit = properties.getOrElse("speedLimit", "").asInstanceOf[String]
+    val pavementClass = properties.getOrElse("pavementClass", "").asInstanceOf[String]
+
+    links.foreach { link =>
+      speedLimit match {
+        case "" =>
+        case _ =>
+          speedLimitService.getExistingAssetByRoadLink(link) match {
+            case Some(value) =>
+              val newSpeedLimitAsset = NewLinearAsset(link.linkId, value.startMeasure, value.endMeasure, NumericValue(speedLimit.toInt), properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              speedLimitService.update(value.id, Seq(newSpeedLimitAsset), AwsUser)
+            case None =>
+              val newSpeedLimitAsset = NewLinearAsset(link.linkId, 0, link.length, NumericValue(speedLimit.toInt), properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              speedLimitService.createMultiple(Seq(newSpeedLimitAsset), speedLimit.toInt, AwsUser, vvhClient.roadLinkData.createVVHTimeStamp(), (_, _) => Unit)
+          }
+      }
+
+      pavementClass match {
+        case "" =>
+        case _ =>
+          val pavementClassValue = DynamicValue(DynamicAssetValue(Seq(DynamicProperty("paallysteluokka", "single_choice", required = false, Seq(DynamicPropertyValue(pavementClass.toInt))))))
+
+          val duplicate = pavedRoadService.getPersistedAssetsByLinkIds(PavedRoad.typeId, Seq(link.linkId))
+          duplicate.isEmpty match {
+            case false =>
+              pavedRoadService.update(Seq(duplicate.head.id), pavementClassValue, AwsUser)
+            case true =>
+              val newPavementClassAsset = NewLinearAsset(link.linkId, 0, link.length, pavementClassValue, properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              pavedRoadService.create(Seq(newPavementClassAsset), PavedRoad.typeId, AwsUser)
+          }
+      }
+    }
+  }
+
+  def validateAndInsertDataset(dataset: Dataset): Int = {
+    var featuresWithoutIds = 0
+
+    var datasetStatus = 0
+    val assets = dataset.geoJson("features").asInstanceOf[List[Map[String, Any]]]
+    val roadlinks = dataset.roadlinks
+
+    if (assets.length != dataset.roadlinks.length) {
+      datasetStatus = 1
+    }
+
+    awsDao.insertDataset(dataset.datasetId, write(dataset.geoJson), write(dataset.roadlinks), datasetStatus)
+    if (datasetStatus != 1) {
+      var roadlinksCount = 0
+
+      assets.foreach(feature => {
+        val featureRoadlinks = roadlinks(roadlinksCount)
+        roadlinksCount += 1
+
+        val properties = feature.getOrElse("properties", "").asInstanceOf[Map[String, Any]]
+        properties.get("id") match {
```

Comment:
Avoid exception.

You can use something like this:

val properties = feature.get("properties").asInstanceOf[Map[String, Any]]

properties match {
Some(value) => //continue validation
case None => //update feature status with the specific error
}